pkg/spoke/hubclientcert/csr.go

```
@@ -0,0 +1,230 @@
+	}
+	for _, v := range usages {
+		write([]byte(v))
+	}
+
+	return fmt.Sprintf("cluster-csr-%s", encode(hash.Sum(nil))), nil
+}
+
+// RequestCertificate will either use an existing (if this process has run
+// before but not to completion) or create a certificate signing request using the
+// PEM encoded CSR and send it to API server, then it will watch the object's
+// status, once approved by API server, it will return the API server's issued
+// certificate (pem-encoded). If there is any errors, or the watch timeouts, it
+// will return an error.
+func RequestCertificate(client certificatesclient.CertificateSigningRequestInterface, csrData []byte, name string, signerName string, usages []certificates.KeyUsage, privateKey interface{}, clusterName string) (req *certificates.CertificateSigningRequest, created bool, err error) {
+	csr := &certificates.CertificateSigningRequest{
+		// Username, UID, Groups will be injected by API server.
+		TypeMeta: metav1.TypeMeta{Kind: "CertificateSigningRequest"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			Labels: map[string]string{
+				clusterNameAnnotation: clusterName,
+			},
+		},
+		Spec: certificates.CertificateSigningRequestSpec{
+			Request:    csrData,
+			Usages:     usages,
+			SignerName: &signerName,
+		},
+	}
+	if len(csr.Name) == 0 {
+		csr.GenerateName = "csr-"
+	}
+
+	req, err = client.Create(context.TODO(), csr, metav1.CreateOptions{})
+	switch {
+	case err == nil:
+	case errors.IsAlreadyExists(err) && len(name) > 0:
+		req, err = client.Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil {
+			return nil, false, formatError("cannot retrieve certificate signing request: %v", err)
+		}
+		if err := ensureCompatible(req, csr, privateKey); err != nil {
+			return nil, false, fmt.Errorf("csr for this cluster already exists but is not valid: %v", err)
+		}
+		return req, false, nil
+	default:
+		return nil, false, formatError("cannot create certificate signing request: %v", err)
```

this is hiding the original error type and isn't a clear win over just returning the error as is.  You know which controller it comes from.